File: archive/swift-backend/Sources/App/Models/Transaction.swift
import Fluent
import Vapor

final class Transaction: Model {
    static let schema = "transactions"

    @ID(key: .id) var id: UUID?

    @Parent(key: "account_id") var account: Account

    @Field(key: "plaid_transaction_id") var plaidTransactionID: String

    @OptionalField(key: "category") var category: String?

    @OptionalField(key: "subcategory") var subcategory: String?

    @Field(key: "type") var type: String

    @Field(key: "name") var name: String

    @Field(key: "amount") var amount: Double

    @OptionalField(key: "iso_currency_code") var isoCurrencyCode: String?

    @OptionalField(key: "unofficial_currency_code") var unofficialCurrencyCode: String?

    @Field(key: "date") var date: Date

    @Field(key: "pending") var pending: Bool

    @OptionalField(key: "account_owner") var accountOwner: String?

    @Timestamp(key: "created_at", on: .create) var createdAt: Date?

    @Timestamp(key: "updated_at", on: .update) var updatedAt: Date?

    @Timestamp(key: "deleted_at", on: .delete) var deletedAt: Date?

    init() {}

    init(
        id: UUID? = nil,
        accountID: UUID,
        plaidTransactionID: String,
        category: String?,
        subcategory: String?,
        type: String,
        name: String,
        amount: Double,
        isoCurrencyCode: String?,
        unofficialCurrencyCode: String?,
        date: Date,
        pending: Bool,
        accountOwner: String?
    ) {
        self.id = id
        self.$account.id = accountID
        self.plaidTransactionID = plaidTransactionID
        self.category = category
        self.subcategory = subcategory
        self.type = type
        self.name = name
        self.amount = amount
        self.isoCurrencyCode = isoCurrencyCode
        self.unofficialCurrencyCode = unofficialCurrencyCode
        self.date = date
        self.pending = pending
        self.accountOwner = accountOwner
    }
}

File: finny-worker/transaction/transaction.go
package transaction

import (
	"errors"
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
	"gorm.io/gorm"
)

type Transaction struct {
	ID                     uuid.UUID       `gorm:"type:uuid;default:gen_random_uuid();primaryKey"`
	AccountID              uuid.UUID       `gorm:"type:uuid;not null"`
	PlaidTransactionID     string          `gorm:"type:text;not null;uniqueIndex:uq:transactions.plaid_transaction_id"`
	Category               string          `gorm:"type:text"`
	Subcategory            string          `gorm:"type:text"`
	Type                   string          `gorm:"type:text;not null"`
	Name                   string          `gorm:"type:text;not null"`
	Amount                 decimal.Decimal `gorm:"type:decimal;not null"`
	IsoCurrencyCode        string          `gorm:"type:text"`
	UnofficialCurrencyCode string          `gorm:"type:text"`
	Date                   time.Time       `gorm:"type:date;not null"`
	Pending                bool            `gorm:"type:boolean;not null"`
	AccountOwner           string          `gorm:"type:text"`
	CreatedAt              time.Time       `gorm:"type:timestamp(6) with time zone;not null;default:CURRENT_TIMESTAMP"`
	UpdatedAt              time.Time       `gorm:"type:timestamp(6) with time zone;not null;default:CURRENT_TIMESTAMP"`
	DeletedAt              gorm.DeletedAt  `gorm:"type:timestamp(6) with time zone"`
}

type TransactionRepository struct {
	db *gorm.DB
}

func NewTransactionRepository(db *gorm.DB) *TransactionRepository {
	return &TransactionRepository{
		db: db,
	}
}

type GetTransactionsInput struct {
	AccountID uuid.UUID
	StartDate time.Time
	EndDate   time.Time
}

func (t *TransactionRepository) GetTransactionsByAccountID(input GetTransactionsInput) ([]Transaction, error) {
	var transactions []Transaction
	err := t.db.Where("account_id = ?", input.AccountID).
		Where("date >= ?", input.StartDate).
		Where("date <= ?", input.EndDate).
		Find(&transactions).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return transactions, nil
		}
		return transactions, err
	}
	return transactions, nil
}
